File: checks/python/src/section508/runner.py
from section508.catalog import rule_ids, rules_by_id
from section508.checks import CHECKS
from section508.document import to_root
from section508.result import Failure, Result


def check(source, rules=None, profile=None) -> Result:
    root = to_root(source)
    ids = rule_ids(rules=rules, profile=profile)
    meta = rules_by_id()
    failures = []
    for rule_id in ids:
        checker = CHECKS.get(rule_id)
        if not callable(checker):
            continue
        details = checker(root) or []
        if details:
            rule = meta.get(rule_id) or {}
            failures.append(
                Failure(
                    id=rule_id,
                    title=rule.get("title") or rule_id,
                    details=list(details),
                )
            )
    return Result(ok=not failures, failures=failures)


def assert_ok(source, rules=None, profile=None) -> None:
    result = check(source, rules=rules, profile=profile)
    if not result.ok:
        raise AssertionError(result.report())

File: checks/ruby/lib/section508/runner.rb
# frozen_string_literal: true

module Section508
  module Runner
    CHECKS = {
      "heading-order" => Checks::HeadingOrder,
      "heading-elements" => Checks::HeadingElements,
      "lists" => Checks::Lists,
      "data-tables" => Checks::DataTables,
      "native-control" => Checks::NativeControl,
      "visible-label" => Checks::VisibleLabel,
      "placeholder-not-label" => Checks::PlaceholderNotLabel,
      "informative-image-alt" => Checks::InformativeImageAlt,
      "decorative-image-silent" => Checks::DecorativeImageSilent,
      "functional-image-names-action" => Checks::FunctionalImageNamesAction,
      "link-purpose" => Checks::LinkPurpose,
      "label-in-name" => Checks::LabelInName,
      "input-purpose" => Checks::InputPurpose,
      "language-of-parts" => Checks::LanguageOfParts,
      "error-identification" => Checks::ErrorIdentification,
      "error-suggestion" => Checks::ErrorSuggestion,
      "accessible-name" => Checks::AccessibleName,
      "name-role-value" => Checks::NameRoleValue,
      "images-of-text" => Checks::ImagesOfText,
      "skip-link" => Checks::SkipLink,
      "color-not-only-cue" => Checks::ColorNotOnlyCue,
      "keyboard-operable" => Checks::KeyboardOperable,
      "no-keyboard-trap" => Checks::NoKeyboardTrap,
      "focus-order" => Checks::FocusOrder,
      "no-change-on-focus" => Checks::NoChangeOnFocus,
      "no-change-on-input" => Checks::NoChangeOnInput,
      "audio-control" => Checks::AudioControl,
      "status-messages" => Checks::StatusMessages,
      "consistent-identification" => Checks::ConsistentIdentification,
      "meaningful-sequence" => Checks::MeaningfulSequence,
      "pointer-cancellation" => Checks::PointerCancellation
    }.freeze

    module_function

    def run(root, ids)
      failures = []
      ids.each do |id|
        checker = CHECKS[id]
        next unless checker.respond_to?(:call)

        details = Array(checker.call(root))
        next if details.empty?

        meta = Catalog.by_id[id] || {}
        failures << Failure.new(
          id: id,
          title: meta["title"] || id,
          details: details
        )
      end
      Result.new(failures)
    end
  end
end
